adoc/chapters/device_compiler.adoc

```diff
@@ -216,132 +217,25 @@ The restriction waiver in <<discarded-statement>> or
+device code is implementation-defined.
 
-The standard {cpp} fixed width types, e.g. [code]#int8_t#, [code]#int16_t#,
-[code]#int32_t#,[code]#int64_t#, should have the same size as defined by the
-{cpp} standard for host and device.
-
-
-[[table.types.fundamental]]
-.Fundamental data types supported by SYCL
-[width="100%",options="header",separator="@",cols="65%,35%"]
-|====
-@ Fundamental data type @ Description
-a@
-[source]
-----
-bool
-----
-   a@ A conditional data type which can be either true or false. The value
-      true expands to the integer constant 1 and the value false expands to the
-      integer constant 0.
-
-a@
-[source]
-----
-char
-----
-   a@ A signed or unsigned 8-bit integer, as defined by the {cpp} core language
-
-a@
-[source]
-----
-signed char
-----
-   a@ A signed 8-bit integer, as defined by the {cpp} core language
-
-a@
-[source]
-----
-unsigned char
-----
-   a@ An unsigned 8-bit integer, as defined by the {cpp} core language
-
-a@
-[source]
-----
-short int
-----
-   a@ A signed integer of at least 16-bits, as defined by the {cpp} core language
-
-a@
-[source]
-----
-unsigned short int
-----
-   a@ An unsigned integer of at least 16-bits, as defined by the {cpp} core language
-
-a@
-[source]
-----
-int
-----
-   a@ A signed integer of at least 16-bits, as defined by the {cpp} core language
-
-a@
-[source]
-----
-unsigned int
-----
-   a@ An unsigned integer of at least 16-bits, as defined by the {cpp} core language
-
-a@
-[source]
-----
-long int
-----
-   a@ A signed integer of at least 32-bits, as defined by the {cpp} core language
-
-a@
-[source]
-----
-unsigned long int
-----
-   a@ An unsigned integer of at least 32-bits, as defined by the {cpp} core language
-
-a@
-[source]
-----
-long long int
-----
-   a@ An integer of at least 64-bits, as defined by the {cpp} core language
-
-a@
-[source]
-----
-unsigned long long int
-----
-   a@ An unsigned integer of at least 64-bits, as defined by the {cpp} core language
-
-a@
-[source]
-----
-float
-----
-   a@ A 32-bit floating-point. The float data type must conform to the IEEE 754
-      single precision storage format.
-
-a@
-[source]
-----
-double
-----
-   a@ A 64-bit floating-point. The double data type must conform to the IEEE 754
-      double precision storage format.  This type is only supported on devices
-      that have [code]#aspect::fp64#.
-
-|====
+The availability of a type in device code does not guarantee that it will be
+supported by all devices.
+Some types are only supported on devices with specific device aspects, as
+described in <<sec:device-aspects>>.
 
+All types which are available in device code must have the same size, alignment
+requirements, and representation on both host and device.
 
+For enumerations with a fixed underlying type, the same fixed type must be
+specified for host and device.
+For enumerations without a fixed underlying type, the implementation must select
+the same underlying type on host and device.
```

Comment:
This second sentence is directed at implementors of the spec.  I think this is redundant.  We already say that "All types which are available in device code" must have the same size and representation.  This statement covers enumerations also.

Reply:
That statement partially covers enumerations, but a different underlying type would still be observable via [`std::underlying_type`](https://en.cppreference.com/w/cpp/types/underlying_type). What is gained by not requiring implementations to use consistent underlying types? Motivation for UB allowances should be strong.

Reply:
OK.  I was not aware of [std::underlying_type](https://en.cppreference.com/w/cpp/types/underlying_type), so I agree this statement is useful.